src/index.c

```diff
@@ -0,0 +1,129 @@
+#include <stdio.h>
+#include <stdlib.h>
+#include <stdbool.h>
+#include "index.h"
+
+
+/* Call to create a new empty Index() */
+Index* init_index(int size) {
+	Index *index = (Index*) malloc(sizeof(Index));
+	if (index == NULL) {
+		printf("Failed to allocate memory for index\n");
+		return NULL;
+	}
+
+	// create contents and set all pointers to null
+	Entry **contents = (Entry **) malloc(size * sizeof(Entry*));
+	for (int i = 0; i < size; i++) {
+		*(contents + i) = NULL;
+	}
+
+	if (contents == NULL) {
+		printf("Failed to allocate memory for index contents.\n");
+		free(index);
+		return NULL;
+	}
+
+	index->size = size;
+	index->positions_filled = 0;
+	index->contents = contents;
+	return index;
+}
+
+/* Inserts a new entry into hash table; where there are
+ * collisions, this program uses chaining. */
+int index_insert(Index *index, int key, char *value) {
+
+	int position = hash(key, index->size);
+
+	Entry *new_entry = (Entry*) malloc(sizeof(Entry));
```

Comment:
Didn't we already allocate memory for the `Entry` when the index was initialized?

Reply:
No, we allocated memory for a list of pointers to `Entry` structs.